Replace WDIO with Cypress

HarelM opened the pull request:
This will move all the e2e test code from using WDIO to using cypress.
It also changes the e2e tests to be in typescript.

HarelM commented:
cc: @nyurik.
This is a second step to improve the e2e tests.
Let me know if you prefer it as a single PR upstream or only after the previous PR is merged.
Generally speaking, both PR are aimed at improving readability and maintainability of e2e tests.

Merging this will update the already open PR upstream, so I'm waiting for an answer from you before merging this to this fork's master.

Let me know how you want to play this.

nyurik commented on .github/workflows/ci.yml:
```diff
@@ -141,52 +141,23 @@ jobs:
+        with:
+          build: npm run build
+          start: npm run start
+          browser: ${{ matrix.browser }}
```
please set your IDE to auto-add trailing newline

nyurik commented on config/webpack.production.config.js:
```diff
@@ -6,16 +6,14 @@ var HtmlWebpackInlineSVGPlugin = require('html-webpack-inline-svg-plugin');
   },
   output: {
-    path: OUTPATH,
+    path: path.join(__dirname, '..', 'build' , 'build'),
```
Shouldn't some lint catch this?

```suggestion
    path: path.join(__dirname, '..', 'build', 'build'),
```

nyurik commented on config/webpack.profiling.config.js:
```diff
@@ -1,13 +1,11 @@
   output: {
     ...webpackProdConfig.output,
-    path: OUTPATH,
+    path: path.join(__dirname, '..', 'build' , 'profiling'),
```
```suggestion
    path: path.join(__dirname, '..', 'build', 'profiling'),
```

nyurik commented on cypress/e2e/driver.ts:
```diff
@@ -0,0 +1,170 @@
+      }
+
+
+}
```
```suggestion
}

```

nyurik commented on cypress/e2e/map.cy.ts:
```diff
@@ -0,0 +1,25 @@
+        //driver.getText(".maplibregl-ctrl-zoom") === "Zoom "+(zoomLevel + 1);
+      })
+    })
+  })
```
```suggestion
  })

```

nyurik reviewed (commented):
i did a cursory look over, marked a few tiny lints, but overall seems like a good direction.

Let's split out "simple changes" from a massive PR. When possible, we shouldn't refactor code and tests at the same time - high chance of a mistake. Could you move these to another PR (possibly one, no need to split them up)
  * upgrade to CI script like latest checkout versions, etc
  * upgrade to typescript version
  * anything else that is NOT test-upgrade related

Another aspect that I wasn't clear about - you do a lot of modifications of the tests - do we actually know if the new tests have the same coverage as before? It is nearly impossible to visually inspect the changes with all the file renames going on. Any way to have some certainty in this process?

Thanks!

HarelM commented:
Once a linter will properly go over the files these will be fixed automatically, so I don't take the time to fix lint issues, there's a need to simply update to latest eslint... Out of scope to this PR.

I can't upgrade to typescript easily, so it's not a step I want to break.

If the code was in a better shape it might have been possible.

The tests cover very little, unfortunately, and there's tons of to-dos in them.

I can't be certain it covers what it used to cover, I have read most of the tests to understand what they do and made sure they pass, I also added back skipped tests.

Most of the work that I did here was to take the original index.js file, copy it to a new ts file with the folder name and added the relevant driver code so that the test is doing the same.
When this wasn't enough I fixed the test and driver code.

I don't know how to break this PR without a rework, which I'm not about to do, sorry...

nyurik commented:
@HarelM my goal is to have some sort of verification for this PR. At the moment, looking at the diff screen is pointless - it shows +3k and -1.5k line changes, with most of them in different files.  Would it be possible to at least keep the filenames unchanged so that they can be visually compared?

HarelM commented:
I've migrated from javascript to typescript, I don't think changing the file names (as I still want to keep the ts extention) will help git think they are the same.

I would recommend to review it as new code instead of trying to compare it to the old code.
I can make a test of changing a single file name to see if git picks it up, but I don't have high hopes.

The main problem is that cypress doesn't use async await and WDIO does, which changes the files dramatically.

HarelM commented:
You can compare `cypress2` branch to master (in this repo).
I moved a file under `cypress...map/index.ts` to see if git "thinks" it's the same file as it was under `test...map/index.js`, but it doesn't - it's too different for it I believe to show them side by side.
So I don't have a good idea how to allow comparing the files with ease.
Let me know how you would like to continue.
